Whisper.net/WhisperProcessor.cs

```diff
@@ -92,23 +93,80 @@ public unsafe (string? language, float probability) DetectLanguageWithProbabilit
             }
         }
 
-        public unsafe WhisperProcessResult Process(float[] samples)
+        public void Process(Stream waveStream)
+        {
+            var waveParser = new WaveParser(waveStream);
+
+            var samples = waveParser.GetAvgSamples();
+
+            Process(samples);
+        }
+
+        public unsafe void Process(float[] samples)
         {
-            segmentIndex = 0;
             fixed (float* pData = samples)
             {
                 NativeMethods.whisper_full(currentWhisperContext, whisperParams, (IntPtr)pData, samples.Length);
-                return new WhisperProcessResult();
             }
         }
 
-        public WhisperProcessResult Process(Stream waveStream)
+        public IAsyncEnumerable<OnSegmentEventArgs> ProcessAsync(Stream waveStream, CancellationToken cancellationToken)
         {
             var waveParser = new WaveParser(waveStream);
-
             var samples = waveParser.GetAvgSamples();
+            return ProcessAsync(samples, cancellationToken);
+        }
+
+        private async IAsyncEnumerable<OnSegmentEventArgs> ProcessAsync(float[] samples, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            ManualResetEventSlim manualResetEventSlim = new();
+            var buffer = new ConcurrentBag<OnSegmentEventArgs>();
+
+            void OnSegmentHandler(object sender, OnSegmentEventArgs @event)
+            {
+                buffer.Add(@event);
+                manualResetEventSlim.Set();

```

Comment:
Shouldn't  manualResetEventSlim` be reset as well?
It seems it will be blocked in set state and the while will become a spinlock.

Reply:
You are right, I missed it.

Reply:
I decided to replace it with some nicer sync, unfortunately .net does not have anything nice built in that works well with async.